accounts/abi/bind/dep_tree_test.go

```diff
@@ -0,0 +1,285 @@
+
+	overrides map[string]common.Address
+}
+
+func makeLinkTestCase(input map[rune][]rune, overrides map[rune]common.Address) *linkTestCase {
+	codes := make(map[string]string)
+	libCodes := make(map[string]string)
+	contractCodes := make(map[string]string)
+
+	inputMap := make(map[rune]map[rune]struct{})
+	// set of solidity patterns for all contracts that are known to be libraries
+	libs := make(map[string]struct{})
+
+	// map of test contract id (rune) to the solidity library pattern (hash of that rune)
+	patternMap := map[rune]string{}
+
+	for contract, deps := range input {
+		inputMap[contract] = make(map[rune]struct{})
+		if _, ok := patternMap[contract]; !ok {
+			patternMap[contract] = crypto.Keccak256Hash([]byte(string(contract))).String()[2:36]
+		}
+
+		for _, dep := range deps {
+			if _, ok := patternMap[dep]; !ok {
+				patternMap[dep] = crypto.Keccak256Hash([]byte(string(dep))).String()[2:36]
+			}
+			codes[patternMap[contract]] = codes[patternMap[contract]] + fmt.Sprintf("__$%s$__", patternMap[dep])
+			inputMap[contract][dep] = struct{}{}
+			libs[patternMap[dep]] = struct{}{}
+		}
+	}
+	overridesPatterns := make(map[string]common.Address)
+	for contractId, overrideAddr := range overrides {
+		pattern := crypto.Keccak256Hash([]byte(string(contractId))).String()[2:36]
+		overridesPatterns[pattern] = overrideAddr
+	}
+
+	for _, pattern := range patternMap {
+		if _, ok := libs[pattern]; ok {
+			// if the library didn't depend on others, give it some dummy code to not bork deployment logic down-the-line
+			if len(codes[pattern]) == 0 {
+				libCodes[pattern] = "ff"
+			} else {
+				libCodes[pattern] = codes[pattern]
+			}
+		} else {
+			contractCodes[pattern] = codes[pattern]
+		}
+	}
+
+	return &linkTestCase{
+		libCodes,
+		contractCodes,
+		overridesPatterns,
+	}
+}
+
+var testKey, _ = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+
+type linkTestCaseInput struct {
+	input          map[rune][]rune
+	overrides      map[rune]struct{}
+	expectDeployed map[rune]struct{}
+}
+
+func testLinkCase(t *testing.T, tcInput linkTestCaseInput) {
+	testAddr := crypto.PubkeyToAddress(testKey.PublicKey)
+	var testAddrNonce uint64
+	overridesAddrs := make(map[common.Address]struct{})
+
+	// generate deterministic addresses for the override set.
+	rand.Seed(42)
+	overrideAddrs := make(map[rune]common.Address)
+	for contract, _ := range tcInput.overrides {
+		var addr common.Address
+		rand.Read(addr[:])
+		overrideAddrs[contract] = addr
+		overridesAddrs[addr] = struct{}{}
+	}
+
+	tc := makeLinkTestCase(tcInput.input, overrideAddrs)
+	allContracts := make(map[rune]struct{})
+
+	for contract, deps := range tcInput.input {
+		allContracts[contract] = struct{}{}
+		for _, dep := range deps {
+			allContracts[dep] = struct{}{}
+		}
+	}
+
+	mockDeploy := func(input []byte, deployer []byte) (common.Address, *types.Transaction, error) {
+		contractAddr := crypto.CreateAddress(testAddr, testAddrNonce)
```

Comment:
I tested this: 
```golang
		fmt.Printf("input is :%x\n", input)
```
It only ever spat out that the input was empty. The way you have decoupled the deployer function as a standalone thing is very good for testability. Please make sure to take advantage of it

Reply:
Right so the input is tested in `v2/lib_test.go` (`TestDeploymentLibraries` is an example).  Omitting `input` causes some tests to fail.

Reply:
Oh yeah, good point that the `Hex2Bytes` method just silently fails if the supplied hex isn't proper.